File: frontend/src/services/catalog/book-copies.ts
import { api } from "@/services/api";
import type { BookCopy, BookCopyCreate, BookCopyUpdate } from "@/types/catalog";

export interface BookCopyListParams {
  book_id?: string;
  status?: BookCopy["status"];
}

export async function listBookCopies(params?: BookCopyListParams): Promise<BookCopy[]> {
  const response = await api.get<BookCopy[]>("/book-copies", { params });
  return response.data;
}

export async function getBookCopy(id: string): Promise<BookCopy> {
  const response = await api.get<BookCopy>(`/book-copies/${id}`);
  return response.data;
}

export async function createBookCopy(payload: BookCopyCreate): Promise<BookCopy> {
  const response = await api.post<BookCopy>("/book-copies", payload);
  return response.data;
}

export async function updateBookCopy(id: string, payload: BookCopyUpdate): Promise<BookCopy> {
  const response = await api.put<BookCopy>(`/book-copies/${id}`, payload);
  return response.data;
}

export async function getBookCopyQr(id: string): Promise<Blob> {
  const response = await api.get<Blob>(`/book-copies/${id}/qr`, {
    responseType: "blob",
  });
  return response.data;
}

File: backend/app/api/v1/endpoints/book_copies.py
"""Book copy endpoints."""

from uuid import UUID

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session

from app.api.deps import get_db, require_roles
from app.models.book_copy import BookCopyStatus
from app.models.user import User
from app.schemas.book_copy import BookCopyCreate, BookCopyResponse, BookCopyUpdate
from app.services.book_copy_service import BookCopyService

router = APIRouter()


def get_book_copy_service(db: Session = Depends(get_db)) -> BookCopyService:
    """Provide a BookCopyService for the current request."""
    return BookCopyService(db)


@router.get("", response_model=list[BookCopyResponse])
def list_book_copies(
    book_id: UUID | None = Query(default=None),
    status_filter: BookCopyStatus | None = Query(default=None, alias="status"),
    _current_user: User = Depends(require_roles("ADMIN", "LIBRARIAN")),
    service: BookCopyService = Depends(get_book_copy_service),
) -> list[BookCopyResponse]:
    """List book copies with optional filters."""
    return service.list_copies(book_id=book_id, status=status_filter)


@router.get("/{copy_id}", response_model=BookCopyResponse)
def get_book_copy(
    copy_id: UUID,
    _current_user: User = Depends(require_roles("ADMIN", "LIBRARIAN")),
    service: BookCopyService = Depends(get_book_copy_service),
) -> BookCopyResponse:
    """Get a book copy by id."""
    return service.get_copy(copy_id)


@router.post("", response_model=BookCopyResponse, status_code=status.HTTP_201_CREATED)
def create_book_copy(
    payload: BookCopyCreate,
    _current_user: User = Depends(require_roles("ADMIN", "LIBRARIAN")),
    service: BookCopyService = Depends(get_book_copy_service),
) -> BookCopyResponse:
    """Create a book copy."""
    return service.create_copy(payload)


@router.put("/{copy_id}", response_model=BookCopyResponse)
def update_book_copy(
    copy_id: UUID,
    payload: BookCopyUpdate,
    _current_user: User = Depends(require_roles("ADMIN", "LIBRARIAN")),
    service: BookCopyService = Depends(get_book_copy_service),
) -> BookCopyResponse:
    """Update a book copy."""
    return service.update_copy(copy_id, payload)
